### PR TITLE
Switch to `snark-verifier` Solidity contracts 

### PR DESCRIPTION
**Pull request by nulltea**
This PR switches to regular Sol contracts generated by `snark-verifier` and adds manual instances parcing in Spectre.sol.

> Before we used modifed Yul contract generation and `halo2-solidity-verifier` for automatic instance parcing.

Step gas cost: `345,618` (before `400,526`)
Rotate gas cost: `670,586` (before `766,145 `)

This also removes patches of `snark-verifier`.

Constracrs are moved to `https://github.com/ChainSafe/spectre-contracts` and added here as submodule

### CODE REVIEW

File: lightclient-circuits/src/committee_update_circuit.rs [hunk 1]
```diff
@@ -108,9 +109,16 @@ impl<S: Spec, F: Field> CommitteeUpdateCircuit<S, F> {
             S::SYNC_COMMITTEE_PUBKEYS_ROOT_INDEX,
         )?;
 
+        println!("ffinalized_header_root: {:?}", finalized_header_root.iter().map(|v| v.value()).collect_vec());
```
Comment (ec2) [inline lightclient-circuits/src/committee_update_circuit.rs]: Use log crate? Also theres a typo

File: lightclient-circuits/src/committee_update_circuit.rs [hunk 2]
```diff
@@ -324,7 +323,7 @@ mod tests {
 
         let timer = start_timer!(|| "committee_update mock prover");
         let prover = MockProver::<Fr>::run(K, &circuit, instance).unwrap();
-        prover.assert_satisfied_par();
+        prover.assert_satisfied();
```
Comment (ec2) [inline lightclient-circuits/src/committee_update_circuit.rs:324]: Why's this changed?

File: lightclient-circuits/src/gadget/crypto/sha256_wide/gate.rs [hunk 3]
```diff
@@ -45,6 +42,27 @@ pub struct ShaBitGateManager<F: Field> {
     pub copy_manager: SharedCopyConstraintManager<F>,
 }
 
+/// Witnesses of a sha256 which are necessary to be loaded into halo2-lib.
+#[derive(Clone, Copy, Debug, CopyGetters, Getters)]
+pub struct LoadedSha256<F: Field> {
+    /// bytes_left of the first row of the first round of this keccak_f. This could be used to determine the length of the input.
```
Comment (ec2) [inline lightclient-circuits/src/gadget/crypto/sha256_wide/gate.rs:49]: Whys this commented out?

File: lightclient-circuits/tests/step.rs [hunk 4]
```diff
@@ -1,182 +1,182 @@
-// The Licensed Work is (c) 2023 ChainSafe
-// Code: https://github.com/ChainSafe/Spectre
-// SPDX-License-Identifier: LGPL-3.0-only
-
-use ark_std::{end_timer, start_timer};
```
Comment (ec2) [inline lightclient-circuits/tests/step.rs]: Did you forget to uncomment

File: lightclient-circuits/src/sync_step_circuit.rs [hunk 5]
```diff
@@ -475,11 +475,12 @@ mod tests {
         )
         .unwrap();
 
-        let instance = StepCircuit::<Testnet, Fr>::get_instances(&witness, LIMB_BITS);
+        // let instance = StepCircuit::<Testnet, Fr>::get_instances(&witness, LIMB_BITS);
```
Comment (ec2) [inline lightclient-circuits/src/sync_step_circuit.rs:478]: delet

File: lightclient-circuits/src/committee_update_circuit.rs [hunk 6]
```diff
@@ -189,29 +195,20 @@ impl<S: Spec, F: Field> CommitteeUpdateCircuit<S, F> {
...
 
+        let finalized_header_root_hilo = {
+            let bytes = finalized_header_root.as_ref();
+            let hash_lo = u128::from_le_bytes(bytes[16..].try_into().unwrap());
+            let hash_hi = u128::from_le_bytes(bytes[..16].try_into().unwrap());
+            [hash_lo, hash_hi].map(bn256::Fr::from_u128)
+        };
+
         let instance_vec = iter::once(poseidon_commitment)
-            .chain(ssz_root.as_ref().iter().map(|b| bn256::Fr::from(*b as u64)))
```
Comment (ec2) [inline lightclient-circuits/src/committee_update_circuit.rs:208]: I might be missing something. Why dont we need the SSZ root anymore?

File: lightclient-circuits/src/committee_update_circuit.rs (hunk 1 above)
Comment (nulltea) [reply lightclient-circuits/src/committee_update_circuit.rs]: pushed clean up commit late...

File: lightclient-circuits/src/gadget/crypto/sha256_wide/gate.rs (hunk 3 above)
Comment (nulltea) [reply lightclient-circuits/src/gadget/crypto/sha256_wide/gate.rs:49]: coppied it from fork, `bytes_left` was used there when for dynamic size message, which we don't use here

File: lightclient-circuits/src/committee_update_circuit.rs (hunk 2 above)
Comment (nulltea) [reply lightclient-circuits/src/committee_update_circuit.rs:324]: looks like it was removed from latest halo2-pse, anyway it's only useful when parallel witness assignment is used.

also it's broken in halo2-axiom https://github.com/axiom-crypto/halo2/issues/32

File: lightclient-circuits/src/committee_update_circuit.rs (hunk 6 above)
Comment (nulltea) [reply lightclient-circuits/src/committee_update_circuit.rs:208]: it was redundant, see tg chat with Veridise

**Review by ec2 — APPROVED** [review]

**Pull request merged**